File: client/src/main/java/au/com/codeka/warworlds/client/game/world/ArrayListStarCollection.java
package au.com.codeka.warworlds.client.game.world;

import java.util.ArrayList;

import au.com.codeka.warworlds.common.proto.Star;

/**
 * A {@link StarCollection} that's backed by a simple {@link ArrayList} of stars.
 */
public class ArrayListStarCollection implements StarCollection {
  private ArrayList<Star> stars;

  public ArrayListStarCollection() {
    stars = new ArrayList<>();
  }

  public static ArrayListStarCollection of(Star... stars) {
    ArrayListStarCollection starCollection = new ArrayListStarCollection();
    for (Star star : stars) {
      starCollection.getStars().add(star);
    }
    return starCollection;
  }

  @Override
  public int size() {
    return stars.size();
  }

  @Override
  public Star get(int index) {
    return stars.get(index);
  }

  @Override
  public int indexOf(long starId) {
    for (int i = 0; i < stars.size(); i++) {
      if (stars.get(i).id.equals(starId)) {
        return i;
      }
    }
    return -1;
  }

  @Override
  public boolean notifyStarModified(Star star) {
    int index = indexOf(star.id);
    if (index >= 0) {
      stars.set(index, star);
      return true;
    }
    return false;
  }

  public ArrayList<Star> getStars() {
    return stars;
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/game/world/ArrayListStarCollection.kt
package au.com.codeka.warworlds.client.game.world

import au.com.codeka.warworlds.common.proto.Star
import java.util.*

/**
 * A [StarCollection] that's backed by a simple [ArrayList] of stars.
 */
class ArrayListStarCollection : StarCollection {
  val stars = ArrayList<Star>()

  override fun size(): Int {
    return stars.size
  }

  override fun get(index: Int): Star {
    return stars[index]
  }

  override fun indexOf(starId: Long): Int {
    for (i in stars.indices) {
      if (stars[i].id == starId) {
        return i
      }
    }
    return -1
  }

  override fun notifyStarModified(star: Star): Boolean {
    val index = indexOf(star.id)
    if (index >= 0) {
      stars[index] = star
      return true
    }
    return false
  }

  companion object {
    fun of(vararg stars: Star): ArrayListStarCollection {
      val starCollection = ArrayListStarCollection()
      for (star in stars) {
        starCollection.stars.add(star)
      }
      return starCollection
    }
  }
}